action/protocol/poll/governance_protocol.go

```diff
@@ -312,8 +312,11 @@ func (p *governanceChainCommitteeProtocol) readActiveBlockProducersByEpoch(ctx c
 	epochHeight := rp.GetEpochHeight(epochNum)
 	crypto.SortCandidates(blockProducerList, epochHeight, crypto.CryptoSeed)
 
+	// TODO: kick-out unqualified delegates based on productivity
```

Comment:
I am thinking about soft-kick-out the unqualified delegates, by multiply a float weight in range [0-1), 0 means hard-kick-out. In this way, unqualified delegates will still be in the list but with a lower rank.

Reply:
that's good idea! but this sort is random sort to pick 24 out of 36, so what will we multiply a float with? And does the float variable depend on historic misbehavior data? How about apply that idea on slashing the stakes?

Reply:
If we want to apply soft/hard concept into kick out, we should multiply a float with the weighted votes before we query candidates (when we calculate and merge with native staking), but at that time we don't know the productivity outcome yet